controllers/packetmachine_controller.go

```diff
@@ -559,6 +656,464 @@ func (r *PacketMachineReconciler) reconcileDelete(ctx context.Context, machineSc
+	}
+
+	desiredVXLAN := int32(desiredConfig.VXLAN)
+	desiredVXLANStr := strconv.Itoa(desiredConfig.VXLAN)
+	desiredStateExists := false
+
+	for _, currentVXLAN := range currentAssignments {
+		if currentVXLAN == desiredVXLAN {
+			desiredStateExists = true
+			continue
+		}
+		if err := r.unassignVXLAN(ctx, portID, currentVXLAN); err != nil {
+			return err
+		}
+	}
+
+	if !desiredStateExists {
+		if err := r.assignVXLAN(ctx, portID, desiredVXLANStr); err != nil {
+			return err
+		}
+		log.Info("Port assigned to VXLAN", "port", desiredConfig.PortName, "vxlan", desiredVXLAN)
+	} else {
+		log.Info("Port already assigned to desired VXLAN", "port", desiredConfig.PortName, "vxlan", desiredVXLAN)
+	}
+
+	return nil
+}
+
+// reconcileLayer2AndBonding ensures the port is set to the correct layer2 and bonding configuration
+func (r *PacketMachineReconciler) reconcileLayer2AndBonding(ctx context.Context, portID string, desiredConfig packet.IPAddressCfg) error {
+	port, _, err := r.PacketClient.PortsApi.FindPortById(ctx, portID).Execute()
+	if err != nil {
+		return fmt.Errorf("failed to get port %s: %w", portID, err)
+	}
+	if port == nil {
+		return fmt.Errorf("port %s not found", portID)
+	}
+
+	if desiredConfig.Layer2 {
+		if err := r.setLayer2(ctx, portID, port, desiredConfig.PortName); err != nil {
+			return err
+		}
+	}
+
+	if desiredConfig.Bonded {
+		if err := r.setBonding(ctx, portID, port, desiredConfig.PortName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getCurrentVXLANAssignments fetches the current VXLAN assignments for a port
+func (r *PacketMachineReconciler) getCurrentVXLANAssignments(ctx context.Context, portID string) ([]int32, error) {
+	vlanAssignList, resp, err := r.PacketClient.PortsApi.FindPortVlanAssignments(ctx, portID).Execute()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get port VLAN assignments: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to get port VLAN assignments: unexpected status code %d", resp.StatusCode)
+	}
+	return getPortVXLANAssignments(vlanAssignList), nil
+}
+
+// unassignVXLAN removes a VXLAN assignment from a port
+func (r *PacketMachineReconciler) unassignVXLAN(ctx context.Context, portID string, vxlan int32) error {
+	vxlanStr := strconv.Itoa(int(vxlan))
+	_, resp, err := r.PacketClient.PortsApi.UnassignPort(ctx, portID).PortAssignInput(metal.PortAssignInput{
+		Vnid: &vxlanStr,
+	}).Execute()
+	if err != nil {
+		return fmt.Errorf("failed to unassign port from VXLAN %d: %w", vxlan, err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("failed to unassign port from VXLAN %d: unexpected status code %d", vxlan, resp.StatusCode)
+	}
+	return nil
+}
+
+// assignVXLAN assigns a VXLAN to a port
+func (r *PacketMachineReconciler) assignVXLAN(ctx context.Context, portID, vxlanStr string) error {
+	_, resp, err := r.PacketClient.PortsApi.AssignPort(ctx, portID).PortAssignInput(metal.PortAssignInput{
+		Vnid: &vxlanStr,
```

Comment:
let's check if this is anyway to fetch vxlan id from vlan value - for ease of configuration / usability

Reply:
There are examples here: https://github.com/equinix/terraform-provider-equinix/blob/343c8fc6a8d3674f3067629e146855c1d3d96dbe/internal/resources/metal/port/helpers.go#L145
Either `vlan_ids` or `vxlan_ids` are accepted in the Terraform `equinix_metal_port` resource. The batch port assignment API accepts either format.

https://deploy.equinix.com/developers/api/metal/#tag/Ports/operation/createPortVlanAssignmentBatch

Reply:
`Vnid` in the PortAssignInput will also take either format: https://deploy.equinix.com/developers/api/metal/#tag/Ports/operation/assignPort

In Terraform, we had to be careful in some usage so that Terraform wouldn't get into a bad state where an Equinix Metal API VLAN ID (UUID) could be returned when a VXLAN ID (1234) value was supplied.